File: python/samples/s5_shopbot.py
# SPDX-License-Identifier: MIT
# Sample 5 — ShopBot (all three pillars), the "build this first" finale.
#
#   • cooperates  — a Triage→Orders/Returns→Fraud handoff graph (Sample 1)
#   • acts        — an approval-gated issue_refund tool (Sample 4)
#   • is tested   — a closing behavior eval on the Returns build (Sample 2)
#
# It's Samples 1 + 4 + 2 on a single shared cast, shown as three short acts.
from __future__ import annotations

import asyncio
import os
import sys
from typing import Annotated, Any

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import ui  # noqa: E402
from agent_framework import AgentResponse, Message, tool  # noqa: E402
from agent_framework.orchestrations import HandoffAgentUserRequest, HandoffBuilder  # noqa: E402

from chat import get_client  # noqa: E402


@tool
def lookup_order(order_id: Annotated[str, "The order id, e.g. 1234"]) -> str:
    """Look up an order by id and return a short status line."""
    return f"Order {order_id}: 1x Blue Mug, $18.00 — delivered on Tuesday."


@tool(name="issue_refund", approval_mode="always_require")
def issue_refund_gated(order_id: Annotated[str, "The order id, e.g. 1234"]) -> str:
    """Issue a refund for an order (gated: requires human approval)."""
    return f"Refund of $18.00 issued for order {order_id}."


@tool(name="issue_refund")
def issue_refund_plain(order_id: Annotated[str, "The order id, e.g. 1234"]) -> str:
    """Issue a refund for an order and return a confirmation."""
    return f"Refund of $18.00 issued for order {order_id}."


def _tool_called(result, name: str) -> bool:
    return any(
        getattr(c, "type", None) == "function_call" and getattr(c, "name", None) == name
        for m in result.messages
        for c in m.contents
    )


async def run() -> None:
    ui.banner(
        "Sample 5 · ShopBot — the finale",
        "Cooperate (handoff) · Act (approval-gated refund) · Tested (eval)",
    )

    client = get_client()
    persist = {"require_per_service_call_history_persistence": True}

    # ── Act I — COOPERATE: a handoff graph routes the customer ──────────────────────────
    ui.rule("act I · cooperate — the handoff graph")

    triage = client.as_agent(
        name="Triage",
        instructions="You are the front desk. Route the customer to Orders or Returns and hand off. "
        "Don't answer yourself.",
        description="Routes the customer to Orders or Returns",
        **persist,
    )
    orders = client.as_agent(
        name="Orders",
        instructions="You answer order-status questions. Use lookup_order to get the facts, then reply briefly.",
        description="Handles order-status questions",
        tools=[lookup_order],
        **persist,
    )
    returns = client.as_agent(
        name="Returns",
        instructions="You handle returns and refunds. Use lookup_order to verify the order. Escalate abuse to Fraud.",
        description="Handles returns and refunds",
        tools=[lookup_order],
        **persist,
    )
    fraud = client.as_agent(
        name="Fraud",
        instructions="You handle suspected fraud. Say the case is escalated. This is the end of the line.",
        description="Suspected fraud — terminal",
        **persist,
    )

    workflow = (
        HandoffBuilder(participants=[triage, orders, returns, fraud])
        .with_start_agent(triage)
        .add_handoff(triage, [orders, returns])
        .add_handoff(orders, [triage])
        .add_handoff(returns, [fraud, triage])
        .build()
    )
    id_to_name = {a.id: a.name for a in (triage, orders, returns, fraud)}

    question = "Where is my order #1234?"
    ui.kv("Customer", f'"{question}"')
    pending = await _handle_turn([e async for e in workflow.run(question, stream=True)], id_to_name)
    if pending:  # close the conversation cleanly
        await workflow.run(responses={r.request_id: HandoffAgentUserRequest.terminate() for r in pending})

    # ── Act II — ACT: an approval-gated refund ──────────────────────────────────────────
    ui.rule("act II · act — the approval gate")

    returns_desk = client.as_agent(
        name="Returns",
        instructions="You process refunds. When asked for a refund, call issue_refund with the order id.",
        tools=[lookup_order, issue_refund_gated],
    )

    refund_ask = "My mug arrived smashed — please refund order #1234."
    ui.kv("Customer", f'"{refund_ask}"')

    with ui.thinking("Returns is working…"):
        result = await returns_desk.run(refund_ask)

    new_inputs: list[Any] = [refund_ask]  # seed once; accumulate across approval rounds
    while result.user_input_requests:
        for req in result.user_input_requests:
            print()
            ui.line(f"  ┌─ approval required · {req.function_call.name} ─────────────", ui.WARN)
            ui.line("  └ a refund moves real money — that's why it's gated", ui.WARN)
            ok = ui.confirm(f"approve {req.function_call.name}?")
            ui.success("approved") if ok else ui.failure("denied")
            new_inputs.append(Message("assistant", [req]))
            new_inputs.append(Message("user", [req.to_function_approval_response(ok)]))
        with ui.thinking("continuing…"):
            result = await returns_desk.run(new_inputs)

    ui.agent_line("Returns", (result.text or "").strip())

    # ── Act III — TESTED: closing behavior eval ─────────────────────────────────────────
    ui.rule("act III · tested — the closing eval")

    eval_returns = client.as_agent(
        name="Returns",
        instructions="When asked for a refund, you MUST call issue_refund with the order id.",
        tools=[issue_refund_plain],
    )
    with ui.thinking("running behavior eval…"):
        eval_result = await eval_returns.run("Refund my order #1234.")
    called_refund = _tool_called(eval_result, "issue_refund")
    ui.badge(called_refund, "Returns build issues a refund when asked  (issue_refund called)")

    ui.rule()
    ui.hint(
        "That's the whole story: agents that cooperate, act under human approval, and are "
        "tested for the behavior you care about — one small app, three pillars."
    )
    ui.press_enter()


async def _handle_turn(events, id_to_name: dict[str, str]) -> list:
    pending: list = []
    for event in events:
        if event.type == "handoff_sent":
            src = id_to_name.get(event.data.source, event.data.source)
            tgt = id_to_name.get(event.data.target, event.data.target)
            ui.write("\n  ⇢ ", ui.MUTED)
            ui.write(src, ui.color_for(src))
            ui.write("  hands off to  ", ui.MUTED)
            ui.line(tgt, ui.color_for(tgt))
        elif event.type == "output" and isinstance(event.data, AgentResponse):
            for message in event.data.messages:
                if message.text:
                    ui.agent_line(message.author_name or str(message.role), message.text)
        elif event.type == "request_info" and isinstance(event.data, HandoffAgentUserRequest):
            for message in event.data.agent_response.messages:
                if message.text:
                    ui.agent_line(message.author_name or str(message.role), message.text)
            pending.append(event)
    return pending


if __name__ == "__main__":
    asyncio.run(run())

File: dotnet/MafDemos/Samples/S5_ShopBot.cs
// SPDX-License-Identifier: MIT
// Sample 5 — ShopBot (all three pillars), the "build this first" finale.
//
// One small app that:
//   • cooperates  — a Triage→Orders/Returns→Fraud handoff graph (Sample 1)
//   • acts        — an approval-gated issue_refund tool (Sample 4)
//   • is tested   — a closing behavior eval on the Returns build (Sample 2)
//
// It's Samples 1 + 4 + 2 on a single shared cast, shown as three short acts.

using System.ComponentModel;
using MafDemos.Infrastructure;
using Microsoft.Agents.AI;
using Microsoft.Agents.AI.Workflows;
using Microsoft.Extensions.AI;

namespace MafDemos.Samples;

public static class S5_ShopBot
{
    [Description("Look up an order by id and return a short status line.")]
    private static string LookupOrder([Description("The order id, e.g. 1234")] string orderId)
        => $"Order {orderId}: 1x Blue Mug, $18.00 — delivered on Tuesday.";

    [Description("Issue a refund for an order and return a confirmation.")]
    private static string IssueRefund([Description("The order id, e.g. 1234")] string orderId)
        => $"Refund of $18.00 issued for order {orderId}.";

    public static async Task RunAsync(IChatClient chat)
    {
        Ui.Banner("Sample 5 · ShopBot — the finale",
            "Cooperate (handoff) · Act (approval-gated refund) · Tested (eval)");

        var lookup = AIFunctionFactory.Create(LookupOrder, name: "lookup_order");
        var refund = AIFunctionFactory.Create(IssueRefund, name: "issue_refund");

        // ── Act I — COOPERATE: a handoff graph routes the customer ───────────────────────
        Ui.Rule("act I · cooperate — the handoff graph");

        ChatClientAgent triage = new(chat,
            "You are the front desk. Route the customer to Orders or Returns and hand off. Don't answer yourself.",
            "Triage", "Routes the customer to Orders or Returns");
        // Orders/Returns get the read-only lookup tool inside the workflow (safe, auto-invoked).
        ChatClientAgent orders = new(chat,
            "You answer order-status questions. Use lookup_order to get the facts, then reply briefly.",
            "Orders", "Handles order-status questions", [lookup]);
        ChatClientAgent returns = new(chat,
            "You handle returns and refunds. Use lookup_order to verify the order. Escalate abuse to Fraud.",
            "Returns", "Handles returns and refunds", [lookup]);
        ChatClientAgent fraud = new(chat,
            "You handle suspected fraud. Say the case is escalated. This is the end of the line.",
            "Fraud", "Suspected fraud — terminal");

        Workflow shop = AgentWorkflowBuilder.CreateHandoffBuilderWith(triage)
            .WithHandoffs(triage, [orders, returns])
            .WithHandoff(returns, fraud, handoffReason: "Suspected fraud or abuse.")
            .WithHandoffs([orders, returns], triage)
            .Build();

        const string question = "Where is my order #1234?";
        Ui.KeyValue("Customer", $"\"{question}\"");
        await RunTurnAsync(shop, [new ChatMessage(ChatRole.User, question)]);

        // ── Act II — ACT: an approval-gated refund ───────────────────────────────────────
        Ui.Rule("act II · act — the approval gate");

        var gatedRefund = new ApprovalRequiredAIFunction(refund);
        AIAgent returnsDesk = chat.AsAIAgent(
            name: "Returns",
            instructions: "You process refunds. When asked for a refund, call issue_refund with the order id.",
            tools: [lookup, gatedRefund]);

        const string refundAsk = "My mug arrived smashed — please refund order #1234.";
        Ui.KeyValue("Customer", $"\"{refundAsk}\"");

        AgentSession session = await returnsDesk.CreateSessionAsync();
        AgentResponse resp;
        using (new Ui.Spinner("Returns is working…"))
            resp = await returnsDesk.RunAsync(refundAsk, session);

        var approvals = resp.Messages.SelectMany(m => m.Contents).OfType<ToolApprovalRequestContent>().ToList();
        while (approvals.Count > 0)
        {
            var replies = new List<ChatMessage>();
            foreach (var req in approvals)
            {
                var toolName = (req.ToolCall as FunctionCallContent)?.Name ?? req.ToolCall.CallId;
                Console.WriteLine();
                Ui.Line($"  ┌─ approval required · {toolName} ─────────────", Ui.Warn);
                Ui.Line("  └ a refund moves real money — that's why it's gated", Ui.Warn);
                var ok = Ui.Confirm($"approve {toolName}?");
                if (ok) Ui.Success("approved"); else Ui.Failure("denied");
                replies.Add(new ChatMessage(ChatRole.User, [req.CreateResponse(ok)]));
            }
            using (new Ui.Spinner("continuing…"))
                resp = await returnsDesk.RunAsync(replies, session);
            approvals = resp.Messages.SelectMany(m => m.Contents).OfType<ToolApprovalRequestContent>().ToList();
        }

        Ui.Write("  [Returns] ", Ui.ColorFor("Returns"));
        Ui.Line((resp.Text ?? "").Trim(), ConsoleColor.White);

        // ── Act III — TESTED: closing behavior eval ──────────────────────────────────────
        Ui.Rule("act III · tested — the closing eval");

        AIAgent evalReturns = chat.AsAIAgent(
            name: "Returns",
            instructions: "When asked for a refund, you MUST call issue_refund with the order id.",
            tools: [refund]);

        AgentResponse evalResp;
        using (new Ui.Spinner("running behavior eval…"))
            evalResp = await evalReturns.RunAsync("Refund my order #1234.");

        var calledRefund = evalResp.Messages.SelectMany(m => m.Contents)
            .OfType<FunctionCallContent>()
            .Any(c => c.Name.Equals("issue_refund", StringComparison.OrdinalIgnoreCase));

        Ui.Badge(calledRefund, "Returns build issues a refund when asked  (issue_refund called)");

        Ui.Rule();
        Ui.Hint("That's the whole story: agents that cooperate, act under human approval, and are " +
                "tested for the behavior you care about — one small app, three pillars.");
        Ui.PressAnyKey();
    }

    // Compact streaming turn runner (same robust shape as Sample 1, plus tool-call lines).
    private static async Task RunTurnAsync(Workflow workflow, List<ChatMessage> conversation)
    {
        await using StreamingRun run = await InProcessExecution.RunStreamingAsync(workflow, conversation);
        await run.TrySendMessageAsync(new TurnToken(emitEvents: true));

        var stream = new Ui.AgentStream();
        string? lastSpeaker = null;
        string? handoffPendingFrom = null;

        await foreach (WorkflowEvent evt in run.WatchStreamAsync())
        {
            if (evt is AgentResponseUpdateEvent e)
            {
                var speaker = e.Update.AuthorName is { Length: > 0 } an ? an : e.ExecutorId;

                foreach (var call in e.Update.Contents.OfType<FunctionCallContent>())
                {
                    if (call.Name.StartsWith("handoff_to_", StringComparison.OrdinalIgnoreCase))
                    {
                        handoffPendingFrom = speaker;
                    }
                    else
                    {
                        stream.End();
                        stream.Reset(); // reprint the speaker label after the tool line
                        Ui.Write("  ⚙ ", Ui.Muted);
                        Ui.Line($"{speaker} calls {call.Name}", Ui.Muted);
                    }
                }

                if (speaker != lastSpeaker)
                {
                    if (handoffPendingFrom is { } from && from != speaker)
                    {
                        stream.End();
                        Ui.Write("  ⇢ ", Ui.Muted);
                        Ui.Write(from, Ui.ColorFor(from));
                        Ui.Write("  hands off to  ", Ui.Muted);
                        Ui.Line(speaker, Ui.ColorFor(speaker));
                        handoffPendingFrom = null;
                    }
                    lastSpeaker = speaker;
                }

                stream.Chunk(speaker, e.Update.Text);
            }
            else if (evt is WorkflowOutputEvent)
            {
                stream.End();
                return;
            }
            else if (evt is WorkflowErrorEvent err)
            {
                stream.End();
                Ui.Failure(err.Exception?.Message ?? "workflow error");
                return;
            }
        }
        stream.End();
    }
}
